### PR TITLE
Update nix to 0.28.0

### PR DESCRIPTION
**Pull request by omprakaash**
Updates nix to 0.28.0. Changed the deprecated functions to new recommended ones. Another noticeable change was the write() ~and close()~ function~s~ in nix now take in `OwnedFd` instead of `RawFd`. Included changes for them as well.

### CODE REVIEW

**Comment by utam0k** [comment]
>  Another noticeable change was the write() and close() functions in nix now take in OwnedFd instead of RawFd. Included changes for them as well.

Where can I find this change? I couldn't find the change you mentioned.
https://github.com/nix-rust/nix/blob/master/CHANGELOG.md#changed

**Comment by omprakaash** [comment]
Yeah, I could not find that in the change-log too. The change was done in this PR: https://github.com/nix-rust/nix/pull/2100.

File: crates/libcontainer/src/process/fork.rs
Comment (YJDoc2) [inline crates/libcontainer/src/process/fork.rs:174]: Need to check if this change is ok (should be, just have to confirm), but in that case, we should also remove the comment above, as not applicable anymore.

File: crates/libcontainer/src/process/container_intermediate_process.rs
Comment (zahash) [inline crates/libcontainer/src/process/container_intermediate_process.rs:135]: why not use `unsafe { std::os::fd::OwnedFd::from_raw_fd( exec_notify_fd ) }`
i think `RawFd` is more semantically related to `OwnedFd` instead of `File`

File: crates/libcontainer/src/process/fork.rs
Comment (anti-entropy123) [inline crates/libcontainer/src/process/fork.rs:173]: It seems we don't need `mman::MapFlags::MAP_ANONYMOUS` anymore.

Comment (zahash) [reply crates/libcontainer/src/process/fork.rs:173]: yes. i have also confirmed that `mmap_anonymous` has this line

```rust
pub unsafe fn mmap_anonymous(
    addr: Option<NonZeroUsize>,
    length: NonZeroUsize,
    prot: ProtFlags,
    flags: MapFlags,
) -> Result<NonNull<c_void>> {
    // ...

    let flags = MapFlags::MAP_ANONYMOUS | flags;

    // ...
}
```

File: crates/libcontainer/src/process/container_intermediate_process.rs
Comment (omprakaash) [reply crates/libcontainer/src/process/container_intermediate_process.rs:135]: Oh Yes ! That is much better. Will change it.

File: crates/libcontainer/src/process/fork.rs
Comment (omprakaash) [reply crates/libcontainer/src/process/fork.rs:173]: Yep that is redundant now. Will remove it. Thanks !

**Comment by utam0k** [comment]
@omprakaash It seems CI failed with lint. May I ask you to check it?

**Comment by omprakaash** [comment]
Oh sorry about that, should have fixed the linting issues now.

**Comment by utam0k** [comment]
@omprakaash Thanks!

**Comment by utam0k** [comment]
@zahash @anti-entropy123 @YJDoc2 May I ask you to leave this PR to you?

**Review by zahash — APPROVED** [review]
looks good

File: crates/libcontainer/src/process/container_intermediate_process.rs
Comment (YJDoc2) [reply crates/libcontainer/src/process/container_intermediate_process.rs]: Marking as resolved, as changed accordingly

File: crates/libcontainer/src/container/tenant_builder.rs
Comment (anti-entropy123) [inline crates/libcontainer/src/container/tenant_builder.rs:152]: Since write_end is `OwnedFd`, why you do not close it by `drop(write_end)`? 

```rust
#[stable(feature = "io_safety", since = "1.63.0")]
impl Drop for OwnedFd {
    #[inline]
    fn drop(&mut self) {
        unsafe {
            ...
            let _ = libc::close(self.fd);
            ...
```
https://doc.rust-lang.org/nightly/src/std/os/fd/owned.rs.html#174-186

File: crates/libcontainer/src/process/container_intermediate_process.rs
Comment (YJDoc2) [inline crates/libcontainer/src/process/container_intermediate_process.rs:143]: Previously, as everything was `rawFd`, we would manually close it in the last close call. However, now we are converting it to the `ownedFd`, using it for `write` and then doing another manual `close` call on it. As per docs, `OwnedFd` will attempt to close the underlying file description on `Drop`, so here wouldn't that cause to get called `close` twice on same fd? I think that is not error-ing, as the impl of drop [as seen here](https://doc.rust-lang.org/src/std/os/fd/owned.rs.html#174-186) says (essentially) `we ignore errors so to not accidentally close another valid file desc` . Do we calling close explicitly here can run into the same problem?

Should we instead convert to `ownedFd` for the write call, let its drop take care of closing the fd, and remove the close call altogether?

File: crates/libcontainer/src/container/tenant_builder.rs
Comment (YJDoc2) [inline crates/libcontainer/src/container/tenant_builder.rs:152]: Same question as the other comment, `write_end` as an `OwnedFd` will call close on drop ; so should we not call close here? If not, a better option to force the closing would be manually calling `drop()` on it at this point.

File: crates/libcontainer/src/process/fork.rs
Comment (YJDoc2) [inline crates/libcontainer/src/process/fork.rs:188]: Why is `as_ptr` needs to be added, when the return type of (original) `mmap` and (new) `mmap_anonymous` is same of `Result<NonNull<c_void>>` ?

**Review by YJDoc2 — CHANGES_REQUESTED** [review]
While other changes are ok, I have some points of concern here regarding the mix of `OwnedFd` and `RawFd` . Main point of concern for this PR, is that in several places, we are converting a RawFd to OwnedFd, and then calling close on it manually when the OwnedFd gets closed in its drop. I feel this should be addressed here.

File: crates/libcontainer/src/container/tenant_builder.rs
Comment (YJDoc2) [reply crates/libcontainer/src/container/tenant_builder.rs:152]: You beat me to it :joy: https://github.com/containers/youki/pull/2728#discussion_r1547168975

File: crates/libcontainer/src/process/fork.rs
Comment (YJDoc2) [reply crates/libcontainer/src/process/fork.rs]: Hey, this still needs to be addressed, the function change is valid, so comment should be updated/removed.

File: crates/libcontainer/src/process/fork.rs
Comment (anti-entropy123) [reply crates/libcontainer/src/process/fork.rs:188]: > Why is `as_ptr` needs to be added, when the return type of (original) `mmap` and (new) `mmap_anonymous` is same of `Result<NonNull<c_void>>` ?

Perhaps because with nix version "0.27.1", the mmap function returns a `Result<*mut c_void>`.

Comment (YJDoc2) [reply crates/libcontainer/src/process/fork.rs:188]: oh yes, I was comparing both from 0.28.0, my bad. Thanks!

File: crates/libcontainer/src/process/container_intermediate_process.rs
Comment (omprakaash) [reply crates/libcontainer/src/process/container_intermediate_process.rs:143]: Ah yes good point, I missed this. Will look into this. I think letting the drop automatically close the fd would be good, since OwnedFd has that feature built in.

File: crates/libcontainer/src/container/tenant_builder.rs
Comment (omprakaash) [reply crates/libcontainer/src/container/tenant_builder.rs]: Yes, since drop is automatically called at end of scope, the close() seems to be redundant.

**Comment by YJDoc2** [comment]
Hey, thanks for updating. Two notes :
- there's conflict in lock file, you can recreate it or copy it over from master and run build to fix it
- Can you add explicit `drop` call where there was close before, and comment the reasoning, maybe linking this PR there? That way in future we know for sure that we have taken care of the fds and the reason for calling drop instead of close

**Comment by YJDoc2** [comment]
Hey @omprakaash , it seems there are still some more conflicts here. In case you are busy, is it ok if I resolve them and push? because the conflicts are in cargo* files, not code, it would be pretty straight forward to resolve. Apart from that, the code is lgtm.

**Comment by omprakaash** [comment]
Hey sorry for the late reply. I did not have the chance to look at this and missed your comment. Would be great if you could that. Thank you !

**Comment by yihuaf** [comment]
I will take care of this.

**Comment by yihuaf** [comment]
Merge conflict is resolved. I will look over the review comments over the weekend and address them.

**Comment by yihuaf** [comment]
@YJDoc2 This is ready for review one last time. I have addressed the merge conflicts and went over the reviews. One note, the closing of the pipes (The fds that were converted from `RawFd` to `OwnedFd`) is actually required. We use the pipes as both a way to communicate between processes and as a barrier for synchronization. The finer control to explicitly close is required, instead of letting the scope takes care of the close through `Drop` trait. By closing one end of the pipe, we immediately signal to the other end of the pipe, so that any blocking calls on the pipe will immediately return. The PR already included the change, but I changed the comments to reflect this information. The previous comment implies that the `drop` was there only to maintain the legacy code, but the explicit close actual serves a purpose in this case, not just because we have to manually close due to the use of `RawFd`.

Let me know if there is anything else you would like to make change.

**Review by YJDoc2 — APPROVED** [review]
Code looks good, but unfortunately there's another conflict due to latest release. Approving so that after resolving and @utam0k 's approval, this can get merged. Thank you @omprakaash and @yihuaf for taking care of this :pray:

**Review by utam0k — APPROVED** [review]
Thanks!

**Comment by yihuaf** [comment]
close #2723

**Pull request merged**
